### PR TITLE
feat: add support for element.closest

### PR DESCRIPTION
**Pull request by tedconn**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -345,12 +345,30 @@ function getElementsByName (name, context = document) {
   return getMatchingElementsByName(elementIterator, name)
 }
 
+function closest (selector, context) {
```
Comment (tedconn) [inline src/index.js:348]: @nolanlawson I tried to use `getFirstMatchingAncestor` but passing in `ast.nodes` didn't work, even though it expects an array of nodes...

**Review by caridy — APPROVED** [review]

File: src/index.js (hunk 1 above)
Comment (nolanlawson) [reply src/index.js:348]: @tedconn I looked into it, it's because the top-level `ast.nodes` is for a comma-separated list of selectors, e.g. `.foo, .bar, .baz`. So if it's passed directly in to `matches()` then it fails because we never get into the `if` or `else if` blocks, and just return true.

File: test/closest.spec.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  })
+})
+
+describe(('closest not matching'), () => {
+  const selector = '.something'
+
+  it('light DOM - does not find closest ancestor', () => {
+    withDom(lightSimple1, context => {
+      const element = context.querySelector('.text')
+      assertResultEqual(selector, element.closest(selector), [], false)
```
Comment (nolanlawson) [inline test/closest.spec.js:42]: I checked, and this test actually checks for `undefined`, not `null`, which is what we're supposed to be returning. As it turns out, `querySelector()` has the same problem, so we've been incorrectly returning `undefined` instead of `null` this whole time. 🤦 

I have a fix, I'll push my commit.

**Comment by nolanlawson** [comment]
This should fix both of the issues I found in the PR, @tedconn quick +1 before we merge?

File: src/index.js [hunk 3]
```diff
@@ -320,7 +320,7 @@ function getElementsByTagNameNS (namespaceURI, tagName, context = document) {
 }
 
 function querySelector (selector, context = document) {
-  return query(selector, context, false)
+  return query(selector, context, false) || null
```
Comment (nolanlawson) [inline src/index.js:323]: Guess this should technically be in another PR since we're fixing querySelector, but I think it's fine to put it in this PR

File: src/index.js [hunk 4]
```diff
@@ -88,7 +88,7 @@ function getParent (element) {
 
 function getFirstMatchingAncestor (element, nodes) {
   let ancestor = getParent(element)
-  while (ancestor) {
+  while (ancestor && ancestor.nodeType === 1) {
```
Comment (nolanlawson) [inline src/index.js:91]: Guess we may as well check for `nodeType` here; it's pretty useless to try to run the matching algorithm on `document` or `shadowRoot`.

Comment (nolanlawson) [reply src/index.js:91]: Actually just realized this is useless because `getParent` never returns anything but an element. I'll remove it

File: src/index.js (hunk 3 above)
Comment (nolanlawson) [reply src/index.js]: D'oh, just realized this is not needed. `querySelector()` was correctly returning null all along

**Comment by nolanlawson** [comment]
Fixed merge conflicts ^

**Review by nolanlawson — APPROVED** [review]
